runtime/queries/metricsview_comparision_toplist.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	mv, err := lookupMetricsView(ctx, rt, instanceID, q.MetricsViewName)
+	if err != nil {
+		return err
+	}
+
+	if mv.TimeDimension == "" {
+		return fmt.Errorf("metrics view '%s' does not have a time dimension", q.MetricsViewName)
+	}
+
+	if q.BaseTimeStart == nil || q.BaseTimeEnd == nil || q.ComparisonTimeStart == nil || q.ComparisonTimeEnd == nil {
+		return fmt.Errorf("undefined time range for comparison on '%s' metrics view ", q.MetricsViewName)
+	}
+
+	sql, args, err := q.buildMetricsTopListSQL(mv, olap.Dialect())
+	if err != nil {
+		return fmt.Errorf("error building query: %w", err)
+	}
+
+	rows, err := olap.Execute(ctx, &drivers.Statement{
+		Query:    sql,
+		Args:     args,
+		Priority: priority,
+	})
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	var data []*runtimev1.MetricsViewComparisonRow
+	for rows.Next() {
+		values, err := rows.SliceScan()
+		if err != nil {
+			return err
+		}
+		measureValues := []*runtimev1.MetricsViewComparisonValue{}
+
+		for i, name := range q.MeasureNames {
+			bv, err := pbutil.ToValue(values[1+i*4])
+			if err != nil {
+				return err
+			}
+
+			cv, err := pbutil.ToValue(values[2+i*4])
+			if err != nil {
+				return err
+			}
+
+			da, err := pbutil.ToValue(values[3+i*4])
+			if err != nil {
+				return err
+			}
+
+			dr, err := pbutil.ToValue(values[4+i*4])
+			if err != nil {
+				return err
+			}
+
+			measureValues = append(measureValues, &runtimev1.MetricsViewComparisonValue{
+				MeasureName:     name,
+				BaseValue:       bv,
+				ComparisonValue: cv,
+				DeltaAbs:        da,
+				DeltaRel:        dr,
+			})
+		}
+
+		dv, err := pbutil.ToValue(values[0])
+		if err != nil {
+			return err
+		}
+
+		data = append(data, &runtimev1.MetricsViewComparisonRow{
+			DimensionName:  q.DimensionName,
+			DimensionValue: dv,
+			MeasureValues:  measureValues,
+		})
+	}
+
+	q.Result = &runtimev1.MetricsViewComparisonToplistResponse{
+		Data: data,
+	}
+
+	return nil
+}
+
+func timeRangeClause(start, end *timestamppb.Timestamp, td string, args *[]any) string {
+	var clause string
+	clause += fmt.Sprintf(" AND %s >= ?", td)
+	*args = append(*args, start.AsTime())
+
+	clause += fmt.Sprintf(" AND %s < ?", td)
+	*args = append(*args, end.AsTime())
+
+	return clause
+}
+
+func (q *MetricsViewComparisonToplist) buildMetricsTopListSQL(mv *runtimev1.MetricsView, dialect drivers.Dialect) (string, []any, error) {
+	dimName := safeName(q.DimensionName)
+	selectCols := []string{dimName}
+	finalSelectCols := []string{}
+	measureMap := make(map[string]int)
+	for i, n := range q.MeasureNames {
+		measureMap[n] = i
+		found := false
+		for _, m := range mv.Measures {
+			if m.Name == n {
+				expr := fmt.Sprintf(`%s as "%s"`, m.Expression, m.Name)
+				selectCols = append(selectCols, expr)
+				finalSelectCols = append(
+					finalSelectCols,
+					fmt.Sprintf(
+						"base.%[1]s, comparison.%[1]s, comparison.%[1]s - base.%[1]s, (comparison.%[1]s - base.%[1]s)/base.%[1]s::DOUBLE",
```

Comment:
I think the subtraction, division and cast to `::DOUBLE` will fail for measures that produce non-numeric values (suppose someone, weirdly, has a measure like `ANY_VALUE(name)`). Maybe also for `HUGEINT` or `DECIMAL` in some cases?

We need a way to protect against that case – any ideas? I guess it's preferable to guard against it in-query vs. running an extra query to pre-check measure result types.

Reply:
We cannot do delta on non-numeric measures, I think we should report it's not supported.

Reply:
Timeseries also fail for `ANY(name)`